src/message_filters/__init__.py

```diff
@@ -316,3 +318,122 @@ def add(self, msg, my_queue, my_queue_index=None):
                     del q[t.nanoseconds]
                 break  # fast finish after the synchronization
         self.lock.release()
+
+
+class TimeSequencer(SimpleFilter):
+    """
+    Sequences messages based on the timestamp of their header.
+
+    At construction, the TimeSequencer takes a duration 'delay' which specifies
+    how long to queue up messages to provide a time sequencing over them.
+    As messages arrive, they are sorted according to their timestamps.
+    A callback for a message is never invoked until the messages' timestamp is
+    out of date by at least the delay. However, for all messages which are out of
+    date by at least delay, their callbacks are invoked in temporal order.
+    If a message arrives from a time prior to a message which has already had its
+    callback invoked, it is thrown away.
+    """
+
+    def __init__(
+        self,
+        input_filter: SimpleFilter,
+        delay: Duration | float,
+        update_rate: Duration | float,
+        queue_size: int,
+        node: Node,
+        msg_stamp_attr: str = "header.stamp",
+    ):
+        """
+        Construct a TimeSequencer filter for a subscriber.
+
+        Args:
+        ----
+        input_filter (SimpleFilter): The input filter to connect to.
+            Typically a Subscriber.
+        delay (Duration | float): The delay (in seconds) to wait for
+            messages to arrive before dispatching them.
+        update_rate (Duration | float): The rate at which to check for
+            messages that are ready to be dispatched.
+        queue_size (int): The maximum number of messages to store.
+        node (Node): The node to create the timer on.
+        msg_stamp_attr (str, optional): The attribute to use for retrieving
+            the timestamp from the message. Should point to a
+            builtin_interfaces.msg.Time field. Defaults to "header.stamp".
+
+        """
+        super().__init__()
+        if not isinstance(delay, Duration):
+            delay = Duration(seconds=delay)
+        if not isinstance(update_rate, Duration):
+            update_rate = Duration(seconds=update_rate)
+        self.delay = delay
+        self.update_rate = update_rate
+        self.queue_size = queue_size
+        self.lock = threading.Lock()
+        self.messages = []
+        self.last_time = Time()
+        self.node = node
+        self.msg_stamp_attrs = msg_stamp_attr.split(".")
+        self.update_timer = self.node.create_timer(
+            self.update_rate.nanoseconds / 1e9, self.dispatch
+        )
```

Comment:
Since this spins a timer up, I think it would make sense to also have a corresponding `shutdown` function to clean up resources:

```
def shutdown(self):
        self.update_timer.cancel()
        self.node.destroy_timer(self.update_timer)
```